feat: wezterm as image provider

akthe-at opened the pull request:
Follow Up to #154 
> I have got a working version implemented but who knows how "good" my implementation is. This might be a bit hacky. I have one big of chunk of things that I want to finish but I wasn't sure how I should go about getting plugin "options" to the new canvas subclass. So my major items to finish in my mind are this.
> 
>  
> 
> Let user defined plugin options pass to the canvas which will control how the terminal split is created/positioned

- [x] Finish Typing/Comments
- [x] Fix Infinity Loop on sending image to terminal 
- [x] implement user configuration options from MoltenOptions into the Wezterm API calls so that users can set their split direction and size preference.
- [x] Finish input validation/error handling
- [ ] ~Add Cell output identification echo under or above graphic in terminal~
- [x] Add to project readme/documentation

akthe-at committed:
add wezterm canvas and option to select it as a canvas

akthe-at committed:
add minimal working example of writing to wezterm canvas via imgcat

akthe-at committed:
Merge branch 'main' into wez_dev

akthe-at committed:
trying to implement split on init

akthe-at committed:
WIP but currently functioning wezterm canvas

akthe-at committed:
updated wezterm.nvim api

needs typing/documentation and error handling/input validation. Before I do that I need to figure out how to get the Molten-Nvim plugin options into the canvas?

akthe-at committed:
updated plugin options for use with wezterm canvas

akthe-at committed:
deinit earlier w/ wezterm canvas

akthe-at committed:
send image to split pane without using auto_image plugin options

akthe-at committed:
prep for draft PR, remove print debug

benlubas commented on rplugin/python3/molten/moltenbuffer.py:
```diff
@@ -146,14 +148,19 @@ def open_image_popup(self, silent=False) -> bool:
+                    if isinstance(self.canvas, WeztermCanvas):
+                        self.canvas.send_image(chunk.img_path)
+                        return True
+
```
I don't think I like this this here. Ideally just opening the output window should show the image again if it's been closed.

The `open_image_popup` function should stick to opening images completely outside of the terminal.

akthe-at committed:
cleanup old code

benlubas commented on rplugin/python3/molten/moltenbuffer.py:
```diff
@@ -215,11 +226,16 @@ def tick(self) -> None:
                     self.open_in_browser(silent=True)
                 if self.options.auto_image_popup:
                     self.open_image_popup(silent=True)
+                if isinstance(self.canvas, WeztermCanvas):
```
Similar, I don't think this should have to be here, opening images should just happen in the canvas.

benlubas commented on rplugin/python3/molten/images.py:
```diff
@@ -199,11 +199,68 @@ def remove_image(self, identifier: str) -> None:
+
+    nvim: Nvim
+
+    def __init__(self, nvim: Nvim):
```
As for getting options into this class, you can add it as a param here, and then just pass it along when the class is instantiated. You might need to add it as a param to all the Canvas classes, I'm not sure how strict python is with that.

benlubas commented on lua/load_wezterm_nvim.lua:
```diff
@@ -0,0 +1,54 @@
+wezterm_api.get_current_pane = function()
+  local current_pane_id = wezterm.get_pane_id()
+  return current_pane_id
+end
```
this function can be shortened to 

```lua
wezterm_api.get_current_pane = wezterm.get_pane_id
```

benlubas commented on rplugin/python3/molten/images.py:
```diff
@@ -199,11 +199,68 @@ def remove_image(self, identifier: str) -> None:
+        """Closes the terminal split that was opened with MoltenInit"""
+        self.wezterm_api.close_image_pane(str(self.image_pane).strip())
+
+    def present(self) -> None:
```
these functions should be doing something. There's not much point in extending the `Canvas` class if they're just empty.

There should be a way to fill these in and also remove the code that you wrote that sorta follows the image popup stuff. Image popup is special b/c it needs to exist alongside canvases. This should be it's own canvas.

benlubas reviewed (changes requested):
Left some comments, tl;dr: do more canvas, less image popup.

Overall this looks really cool. I'm happy to have something that will work well for native windows users.

It also doesn't look like you used the black formatter, or if you did, it's not respecting the configuration in pyproject.toml for line length. You don't have to format now, but I'll probably ask you to format the project with black before this is merged. just for consistency.

akthe-at commented:
> Left some comments, tl;dr: do more canvas, less image popup.
> 
> Overall this looks really cool. I'm happy to have something that will work well for native windows users.
> 
> It also doesn't look like you used the black formatter, or if you did, it's not respecting the configuration in pyproject.toml for line length. You don't have to format now, but I'll probably ask you to format the project with black before this is merged. just for consistency.

Really appreciate the quick feedback and think it all makes a lot of sense. 

1.  Forgot to turn off ruff-lsp before saving and it auto formatted...was waiting for legit formatting and that got away from me. 
2. I totally agree about trying to move more into the canvas, I had something kind of working like that but ran into an issue with the processes behind the canvas updating. It seems like there is something checking for updates cyclically for the code cells? When I was trying to figure it out late at night I got tired and gave up.

akthe-at committed:
fix bad formatting

akthe-at committed:
shorten method/api func name

benlubas commented:
The UI gets updated a lot, that's just how it was done in Magma, I've actually reduced the amount of times they're updated quite a bit since forking this 😅.

The way I implemented virtual text images kinda just ignores most of the updates after the cell is done running I think. I think all the updates that happen now are required to have the floating windows reposition on scroll or resize/update as code runs.

I'm a little pressed for time this week, but when I have time, I can do a more thorough writeup of when each canvas function gets called if that would help.

---

Separate tooling question, ruff is black-compatible, right? I think you could just through a ruff configuration into the pyproject.toml and have it format the same way?

If that's the case, feel free to do so, I'm planning to switch over to ruff eventually anyway.

heh, you could say ruff is "blackwards-compatible". I'll see myself out.

akthe-at committed:
fix: accidentally changed internal api name

akthe-at committed:
add options to wezterm canvas, fix split on init setup

idk how i lost this during the creation of the draft PR

akthe-at committed:
add ruff linelength to match black

akthe-at committed:
remove wezterm references that were nested w/ image popup

akthe-at committed:
move actions back inside wezterm canvas for the most part

Works but is bugy. Will send images to the terminal...but will keep going in an inifinity loop Need to figure out how to fix, i think its sending one image with every check to update

akthe-at commented:
> The UI gets updated a lot, that's just how it was done in Magma, I've actually reduced the amount of times they're updated quite a bit since forking this 😅.
> 
> The way I implemented virtual text images kinda just ignores most of the updates after the cell is done running I think. I think all the updates that happen now are required to have the floating windows reposition on scroll or resize/update as code runs.
> 
> I'm a little pressed for time this week, but when I have time, I can do a more thorough writeup of when each canvas function gets called if that would help.
> 
> Separate tooling question, ruff is black-compatible, right? I think you could just through a ruff configuration into the pyproject.toml and have it format the same way?
> 
> If that's the case, feel free to do so, I'm planning to switch over to ruff eventually anyway.
> 
> heh, you could say ruff is "blackwards-compatible". I'll see myself out.

@benlubas Haha yeah they are supposed to be compatible! I just had to add the toml configuration like you mentioned, which I added to the draft PR. I was able to add the options (haven't implemented using them) and I still need to work on fixing the infinity loop bug.

akthe-at committed:
Merge pull request #1 from benlubas/main

catchup with upstream

akthe-at commented:
The biggest issue related to the current "infinite loop" is that when you initially  send a code chunk that has an image output, things are seemingly fine. The problem that I am seeing is when the cursor moves out of the cell and then tries to re-enter. When the output window goes to redraw itself it starts spamming many calls to canvas.add_image().

It continues to loop and send the image so long as the cursor remains in the cell. If you manage to get the cursor out of the cell, it stops.

This behavior does _not_ happen with 
` vim.g.molten_auto_open_output = false `

In fact, with the above setting = false, running the code as a .py, .md, or .ipynb files all work as expected, and no issues! Therefore, I am thinking I need to dive deep into the logic/flow of `MoltenBuffer.update_interface()` and its interactions thereafter?

akthe-at committed:
feat: add in user options to split size/dir

benlubas commented:
It could be that somewhere in your implementation of `canvas.add_image()` you're losing focus of the neovim instance, and that's causing molten to think the user has left the cell, and then you throw focus back to molten, and it thinks you're entering the cell and the output needs to be re-rendered?

akthe-at committed:
Merge branch 'wez_dev' of github.com:akthe-at/molten-nvim into wez_dev

akthe-at commented:
> It could be that somewhere in your implementation of `canvas.add_image()` you're losing focus of the neovim instance, and that's causing molten to think the user has left the cell, and then you throw focus back to molten, and it thinks you're entering the cell and the output needs to be re-rendered?

Oh that is totally it! to send the image the focus moves from the neovim instance, to the wezterm terminal split, and back to the neovim instance. I need to set it up so that an image is only sent to the terminal when rendering the cell the first or manually re running it, not with each entrance into the cell.

akthe-at committed:
add in input validation and type annotations

akthe-at committed:
disallow mixing auto_open_output w/ wezterm

akthe-at committed:
cleanup

akthe-at committed:
this should take care of diff in platforms

akthe-at commented:
So it really seems like there is a lot of interplay between scrolling / entering and exiting cells/ etc. with updating interfaces. Without drastically changing that code and due to the nature of moving between the instance of neovim to the wezterm split and back...it seems like the most feasible fix for this is to not allow auto_open_output to be true when image_provider = "wezterm".  

There are two benefits to this limitation currently. 1, It fixes the infinity loop that I've mentioned before and 2, when you load an ipynb file that has previous outputs, it won't spam the terminal with all of the previous outputs. This will increase loading time and provider a better user experience I think. This would be OK with  a small notebook with only a few outputs but would not be nice if you had dozens of plots ( or more ). Thoughts? 

The other two major milestones to finish would be trying to put a text label under each plot in the terminal that would identify the plot with the cell output. Finally, to add documentation related to this feature to the readme area(s).

benlubas commented:
I'm fine with having this option disable auto show output. I think this should be enforced in code tho, not just mentioned in the readme, bc many people don't read it.

Implementation can look like: add an if statement to the end of the options init function which checks for the wezterm option. If it and auto open are true, first set auto open to false and then warn the user about the bad configuration. Encouraging them to disable auto open if they want to use wezterm image provider.

akthe-at commented:
> I'm fine with having this option disable auto show output. I think this should be enforced in code tho, not just mentioned in the readme, bc many people don't read it.
> 
> Implementation can look like: add an if statement to the end of the options init function which checks for the wezterm option. If it and auto open are true, first set auto open to false and then warn the user about the bad configuration. Encouraging them to disable auto open if they want to use wezterm image provider.

I had implemented it [like this already](https://github.com/benlubas/molten-nvim/pull/162/commits/4aa7550260eabbee266896d48f92d4425fa03180) but I can change it to the way you mentioned above. I don't have a preference/opinion

benlubas commented:
Oh yeah that's perfectly fine too!

akthe-at committed:
change default split dir and size

right side split direction and 40% size seem like better UX defaults

akthe-at committed:
docs: wezterm img docs

added to main readme, windows readme, and added a gif. I've never added a gif to a readme on github before so I am not sure where it goes

benlubas commented:
This doesn't currently work if you're in tmux it looks like. I wonder if there's a way to get it to work, but I guess that would be on willothy's plugin right? 

Just something to warn about in the readme for this configuration option.

benlubas committed:
docs, and other code cleanup

akthe-at commented:
> This doesn't currently work if you're in tmux it looks like. I wonder if there's a way to get it to work, but I guess that would be on willothy's plugin right?
> 
> Just something to warn about in the readme for this configuration option.

Did the pane splitting work and just not sending of the plot? I didn't even think about tmux functionality since I can't use it on non-WSL windows environment. The Wezterm imgcat command does have a tmux flag that I can pass in but I don't see anything related to the pane splits [wezterm docs](https://wezfurlong.org/wezterm/cli/imgcat.html)

benlubas commented:
Hopefully you don't mind, I cleaned up a few things, and added a new section to the readme to better explain the two image providers now. I've tested it out on mac and it works great. The only thing that I notice is that it leaves a blank prompt between each image. 

<img width="598" alt="image" src="https://github.com/benlubas/molten-nvim/assets/56943754/96860a35-0dae-4a82-bdc4-4b05e02d17a7">

Tbh, I haven't looked at all of the code too closely yet, so I still might have more changes to request

benlubas commented on lua/load_wezterm_nvim.lua:
```diff
@@ -0,0 +1,97 @@
+--- @param size number the size to validate
+--- @return number validated size if valid
+local validate_split_size = function(size)
+  if size < 0 or size > 100 or size == nil then
```
nil check needs to come first, otherwise this will throw an error trying to compare nil to a number.

benlubas commented on lua/load_wezterm_nvim.lua:
```diff
@@ -0,0 +1,97 @@
+--- @return number validated size if valid
+local validate_split_size = function(size)
+  if size < 0 or size > 100 or size == nil then
+    vim.api.nvim_err_writeln(
```
Can you use the notify api here? I think that I'd prefer if this was just a warning, b/c we do have a validation and default that we fall back to.

akthe-at replied:
Do you want me to create a separate lua module that has the wrapper functions setup for the notify api like you did on the python side...or can we use pynvim to call that python code from this side (which seems maybe weird because its calling right back to the vim side of things?). I looked at the load_image_nvim.lua for reference initially and didn't see the notify usuage so that is why I chose this route.

akthe-at replied:
@benlubas just wanted to tag you in this question because I know git comments get lost in all the notifications for me sometimes.

benlubas replied:
Sorry, even that ping got lost. You can just use the notify API. If the lua code base grows we can refactor then. Just do this to make sure the messages are consistent:

```lua
vim.notify("[Molten] <your message>", ...)
```

@akthe-at

akthe-at committed:
fix: empty prompt between plot fix?

Can you test ? No mac access to test but this might be it?

benlubas commented on lua/load_wezterm_nvim.lua:
```diff
@@ -0,0 +1,97 @@
+--- @param size number, size of the new pane
+--- @return number image_pane_id the new pane id
+wezterm_api.wezterm_molten_init = function(initial_pane_id, direction, size)
+  direction = "--" .. (validate_split_dir(direction) or "right")
```
`validate_split_dir` already handles the default value, you could remove the `or "right"` here

akthe-at commented:
> Hopefully you don't mind, I cleaned up a few things, and added a new section to the readme to better explain the two image providers now. I've tested it out on mac and it works great. The only thing that I notice is that it leaves a blank prompt between each image.
> 
> <img alt="image" width="598" src="https://private-user-images.githubusercontent.com/56943754/315765894-96860a35-0dae-4a82-bdc4-4b05e02d17a7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEwNjI4NDEsIm5iZiI6MTcxMTA2MjU0MSwicGF0aCI6Ii81Njk0Mzc1NC8zMTU3NjU4OTQtOTY4NjBhMzUtMGRhZS00YTgyLWJkYzQtNGIwNWUwMmQxN2E3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIxVDIzMDkwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJmMmU2YmFmYjcxOGYwNjYxNTU4MmRhNmZjMTAzOGJmMzQ4Njk2YTY4ODYwYzY5NzUxYmMwZGQ3NTg0MDMyY2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0eMHE5PE0OyYIaTKrasS5yrCQ07LtwxnMovmTnnIhj8">
> Tbh, I haven't looked at all of the code too closely yet, so I still might have more changes to request

I don't mind at all! I didn't know what the etiquette was because I don't know how to have your voice for the docs and to keep with your vision of the plugin. I tried one thing to see if it would fix the empty prompt, if that doesnt, removing the /r on that line will probably do it. My prompt was actually overwriting the bottom part of my plot so i was trying to get away from that...it didnt work...and i forgot to go back and remove it.

benlubas commented on docs/Windows.md:
```diff
@@ -91,10 +92,49 @@ Uberzugpp should work on WSL. You might have to configure it in a special way.
+needing an external pop-up window like used with `:MoltenImagePopup`. You can find the instructions
+for downloading and setting up Wezterm [here](https://wezfurlong.org/wezterm/install/windows.html).
+
+![](wezterm.gif)
```
Can you upload this gif to github and then insert it like that so that it's not hosted in the repo. Just so people don't have to download the gif when they clone the repo.

benlubas reviewed (changes requested):
Just a couple really small things.

This is a great contribution, ~~and I see you already pushed a fix for the blank lines on no windows OS's, thanks!~~ ~~oh wait that didn't work yet. Let me try to remove `\r` as you suggest.~~ wait no, it did I forgot to pull b/c I'm dumb lol. thank you

akthe-at committed:
fix: enable tmux passthru of images if present

I believe this will allow the terminal to read if its passing through tmux...if tmux is configured to emit user env vars which i believe is a user config option. I cant test this on my own but passing 'enable' broke sending plots for me on windows, but leaving in 'detect' with the flag does not.

akthe-at commented:
Lots of great comments @benlubas can't wait to address them for you tonight/tomorrow. Appreciate the patience you're showing me, and I'm learning a lot from the feedback...(like duh ?why? didn't I do it that way originally for several of them)

benlubas commented:
Yeah don't worry about taking time, open source is a free-time thing, and not everyone has a ton of it like me (a college student with a very light course load), so don't feel rushed at all!

I'll test out that tmux change, thanks for looking into that.

benlubas commented:
Tmux is still causing the same problem. Here's what happens:
1. run the cell
2. split opens
3. nothing gets typed into the split
4. cell gets folded
5. some text inside the cell is changed

Those last two points make me think there are some escape codes sent to the terminal that tmux is eating instead of letting them go through to wezterm. There is a way around that if it is the issue, but I'm not sure. I'm also thinking this could be a wezterm.nvim bug, not something we would be able to handle.

If you care to look into this that would be awesome, but I'm completely fine with just leaving a "this doesn't work in tmux" in the readme.

akthe-at commented:
I found this in the wezterm docs, can you check on this setting in your config and see if that makes a difference?

I can't get this formatting to work on my phone sorry.

![Screenshot_20240321-212438.png](https://github.com/benlubas/molten-nvim/assets/105152139/f90e434a-802a-4a80-ae47-9f922bd406e6)

akthe-at committed:
fix: clean up some validation logic

benlubas commented:
I still see the same issue even after setting up the shell integration.

akthe-at committed:
fix: add external link to wezterm.gif

akthe-at commented:
That is pretty annoying that it still didn't work but I guess if you can use tmux you might as well use image.nvim ? I do wonder if passing the flag as "enable" would work now that you have done the shell integration? I think the only way we could manage that is if the users had to set if they were using tmux as a plugin option... If it even worked...

I can start preparing the documentation to warn about tmux compatibility if you don't want to keep trying work arounds (I hate that I can't test it myself).

benlubas committed:
Merge branch 'main' into wez_dev

benlubas commented:
Leaving out tmux compatibility is fine for now, or I can test the enable flag tomorrow.

akthe-at committed:
fix: use vim notify api

akthe-at committed:
fix (hopeful): add option to set explicit flag for tmux

Last ditch effort to pass the enable option to the --tmux-passthru flag.
Fingers crossed.

akthe-at commented:
If the enable option works,  I will add more to the documentation! I didn't want to waste anytime adding to that if it was not going to work. Don't forget to set the option to true in your config to test it out :D

benlubas commented:
Still not working in tmux unfortunately.

We can just add to the readme that it won't work, I think that's fine.

akthe-at committed:
Revert "fix (hopeful): add option to set explicit flag for tmux"

This reverts commit 332aa8b5c84559f6090c3b9347bc0cf85b7e13b5.

akthe-at committed:
add warning about tmux issue with wezterm provider

feel free to change this to how you want this information to be
conveyed.

benlubas commented:
@akthe-at is this ready for a final look over? Or are you still looking to check off that last TODO item at the top?

I'm good with either, just want to make sure you're not waiting on me.

akthe-at commented:
@benlubas I was not able to figure out how to get that working to where it looked/worked nicely. I think I would prefer to set that as a future goal if possible.

benlubas committed:
docs: touchup readme

benlubas commented:
Will test this out for a few days before cutting a release. Thank you so much for all the work @akthe-at !

akthe-at commented:
@benlubas I enjoyed getting to work on this with you. Always felt like you were super helpful and welcoming to your project. I look forward to this being part of my daily workflow.
